coderd/database/pubsub.go

```diff
@@ -15,29 +15,174 @@ import (
 // Listener represents a pubsub handler.
 type Listener func(ctx context.Context, message []byte)
 
+// ListenerWithErr represents a pubsub handler that can also receive error
+// indications
+type ListenerWithErr func(ctx context.Context, message []byte, err error)
+
+// ErrDroppedMessages is sent to ListenerWithErr if messages are dropped or
+// might have been dropped.
+var ErrDroppedMessages = xerrors.New("dropped messages")
+
 // Pubsub is a generic interface for broadcasting and receiving messages.
 // Implementors should assume high-availability with the backing implementation.
 type Pubsub interface {
 	Subscribe(event string, listener Listener) (cancel func(), err error)
+	SubscribeWithErr(event string, listener ListenerWithErr) (cancel func(), err error)
 	Publish(event string, message []byte) error
 	Close() error
 }
 
+// msgOrErr either contains a message or an error
+type msgOrErr struct {
+	msg []byte
+	err error
+}
```

Comment:
Thoughts on moving the in-memory pubsub to its own package? I feel like I made a mistake bundling it inside `database` when I created it, and this seems like an opportunity to fix that.

Reply:
Happy to move it, but probably will open a separate PR.  Where do you think it should go?  subpackage of database?  

We could move the interface & in-memory pubsub to `coderd/pubsub`?